packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * WordPress dependencies
+ */
+import { getScrollContainer } from '@wordpress/dom';
+import { useLayoutEffect } from '@wordpress/element';
+
+export default function useListViewScrollIntoView( {
+	isSelected,
+	selectedClientIds,
+	rowItemRef,
+} ) {
+	const numBlocksSelected = selectedClientIds?.length;
+
+	useLayoutEffect( () => {
+		// Skip scrolling into view if more than one block is selected.
+		// This is to avoid scrolling the view in a multi selection where the user
+		// has intentionally selected multiple blocks within the list view,
+		// but the initially selected block may be out of view.
+		if (
+			! isSelected ||
+			numBlocksSelected !== 1 ||
+			! rowItemRef?.current ||
+			! rowItemRef.current.scrollIntoView
+		) {
+			return;
+		}
+
+		const scrollContainer = getScrollContainer( rowItemRef.current );
+		const { ownerDocument } = rowItemRef.current;
+		const { defaultView } = ownerDocument;
+
+		const windowScroll =
+			scrollContainer === ownerDocument.body ||
+			scrollContainer === ownerDocument.documentElement;
+
+		// If the there is no scroll container, of if the scroll container is the window,
+		// do not scroll into view, as the block is already in view.
+		if ( windowScroll || ! scrollContainer ) {
+			return;
+		}
+
+		const { top, height } = rowItemRef.current.getBoundingClientRect();
+		const topOfScrollContainer =
+			scrollContainer.getBoundingClientRect().top;
+
+		const { innerHeight } = defaultView;
+
+		// If the selected block is not currently visible, scroll to it.
+		if ( top < topOfScrollContainer || top + height > innerHeight ) {
```

Comment:
This works for the list view, but I think it is by accident. Instead of checking for the window's inner height, we should check for the `scrollContainer`'s height.

The current implementation also has a bit of inconsistency. If you click on a row that has its top half covered by the container, then the row will be scrolled to show the full height in the viewport. This is a nice improvement, and I think it should be the expected behavior. However, if the same happens for a row that has its bottom half covered by the container, clicking it won't scroll the container.

Perhaps, we can refactor this into something like this for simplicity:

```js
const rowRect = rowItemRef.current.getBoundingClientRect();
const scrollContainerRect = scrollContainer.getBoundingClientRect();

if ( rowRect.top < scrollContainerRect.top || rowRect.bottom > scrollContainerRect.bottom ) {
  rowItemRef.current.scrollIntoView();
}
```

Reply:
Ah, yes, that logic's much better, thank you! I think my previous logic was a bit of a mix between two different approaches I'd hacked together but that's much neater and more reliable 🙂